File: 백준/Gold/1987. 알파벳/알파벳.cc
#include <bits/stdc++.h>

#define FASTIO ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0)

using namespace std;

int r, c;

int DX[4] = {-1, 1, 0, 0};
int DY[4] = {0, 0, -1, 1};

vector<vector<char>> board;

// 말은 상하좌우로 인접한 네 칸 중 한 칸으로 이동
// 같은 알파벳이 적힌 칸을 두 번 지날 수 없다. => 알파벳에 대한 방문 체크 필요
// 알파벳은 모두 26개이므로 배열로 해결 가능
vector<bool> visited;

void input() {
  FASTIO;
  cin >> r >> c;
  board.resize(r + 1, vector<char>(c + 1));
  visited.resize('Z' - 'A' + 1, false);

  for (int i = 1; i <= r; i++) {
    for (int j = 1; j <= c; j++) {
      cin >> board[i][j];
    }
  }
}

bool isValidRange(int x, int y) { return 1 <= x && x <= r && 1 <= y && y <= c; }

int answer = -1;

// 좌측 상단에서 부터 시작해서, 말이 최대한 몇 칸을 지날 수 있는가?
// (x, y)에 있는 말, 현재 count 개의 칸을 지남
// 꼭 도착을 하지 않아도 된다!
void solve(int x, int y, int count) {
  answer = max(answer, count);

  for (int dir = 0; dir < 4; dir++) {
    int nx = x + DX[dir];
    int ny = y + DY[dir];

    // 다음 칸으로 갈 수 있는가?
    if (!isValidRange(nx, ny)) continue;
    
    // 이미 방문한 알파벳인가?
    int next_alpha_idx = board[nx][ny] - 'A';
    if (visited[next_alpha_idx]) continue;
   
    visited[next_alpha_idx] = true;
    solve(nx, ny, count + 1);
    visited[next_alpha_idx] = false;
  }
}

int main() {
  input();

  visited[board[1][1] - 'A'] = true;
  solve(1, 1, 1);

  cout << answer;

  return 0;
}

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static FastReader scan = new FastReader();

	static final int[] DX = {-1, 1, 0, 0};
	static final int[] DY = {0, 0, -1, 1};
	static int r, c;
	static char[][] board;
	static boolean[] visited;

	static int answer = -1;

	public static void main(String[] args) {
		init();
		// solve();

		visited[board[0][0] - 'A'] = true;
		dfs(0, 0, 1);

		System.out.println(answer);
	}

	static void dfs(int x, int y, int count) {
		answer = Math.max(answer, count);

		// recursive case
		for (int d = 0; d < 4; d++) {
			int nx = x + DX[d];
			int ny = y + DY[d];

			// 더 이상 진행할 수 없는 경우
			if (isOutOfRange(nx, ny) || visited[board[nx][ny] - 'A']) {
				continue;
			}

			visited[board[nx][ny] - 'A'] = true;
			dfs(nx, ny, count + 1);
			visited[board[nx][ny] - 'A'] = false;
		}
	}

	static boolean isOutOfRange(int x, int y) {
		return x < 0 || x >= r || y < 0 || y >= c;
	}

	static void init() {
		r = scan.nextInt();
		c = scan.nextInt();
		board = new char[r][c];
		for (int i = 0; i < r; i++) {
			board[i] = scan.next().toCharArray();
		}

		visited = new boolean['Z' - 'A' + 1];
	}

	static class FastReader {
		BufferedReader br;
		StringTokenizer st;

		public FastReader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}

		public FastReader(String fileName) {
			try {
				br = new BufferedReader(new FileReader(fileName));
			} catch (IOException e) {
				e.printStackTrace();
			}
		}

		String next() {
			while (st == null || !st.hasMoreElements()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
			return st.nextToken();
		}

		int nextInt() {
			return Integer.parseInt(next());
		}
	}
}
